### ISSUE TITLE
tainting inconsistency with . operator

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6525]
Migrated from [rt.perl.org#22270](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=22270) (status was 'resolved')

Searchable as RT22270$

**Comment by p5pRT** [comment]
### From nick@penfold.uucp

#### Created by nick@penfold.uucp



$ perl5\.00405 \-T \-le 'delete $ENV\{PATH\}; my $d = $^X \. "" \. \`/bin/true\`'
Insecure dependency in \`\` while running with \-T switch at \-e line 1\.
$ perl5\.00405 \-T \-le 'delete $ENV\{PATH\}; my $d = "" \. $^X \. \`/bin/true\`'
Insecure dependency in \`\` while running with \-T switch at \-e line 1\.
$ perl5\.00405 \-T \-le 'delete $ENV\{PATH\}; my $d = lc \($^X\) \. \`/bin/true\`'
Insecure dependency in \`\` while running with \-T switch at \-e line 1\.
$ perl5\.00405 \-T \-le 'delete $ENV\{PATH\}; my $d = $^X \. \`/bin/true\`'
$

I don't see why that last is any different\. Surely it should also be
an insecure dependency?

Bug still present in blead

Nicholas Clark

<details><summary>Perl Info</summary>

```


Site configuration information for perl 5.00405:

Configured by nick at Tue Jan  7 14:56:55 GMT 2003.

Summary of my perl5 (5.0 patchlevel 4 subversion 5) configuration:
  Platform:
    osname=linux, osvers=2.4.19, archname=i686-linux-gcov
    uname='linux penfold 2.4.19 #1 sat aug 10 22:59:35 bst 2002 i686 unknown unknown gnulinux '
    hint=recommended, useposix=true, d_sigaction=define
    bincompat3=n useperlio=undef d_sfio=undef
  Compiler:
    cc='gcc-3.2', optimize='-O3 -fprofile-arcs', gccversion=3.2.1
    cppflags='-Dbool=char -DHAS_BOOL -I/usr/local/include'
    ccflags ='-Dbool=char -DHAS_BOOL -I/usr/local/include'
    stdchar='char', d_stdstdio=undef, usevfork=false
    intsize=4, longsize=4, ptrsize=undef, doublesize=undef
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='gcc-3.2', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldbm -ldl -lm -lc -lcrypt
    libc=/lib/libc-2.3.1.so, so=so
    useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl 5.00405:
    /home/nick/lib/i686-linux-gcov/5.00405
    /home/nick/lib
    /home/nick/lib/site_perl/i686-linux-gcov
    /home/nick/lib/site_perl
    .


Environment for perl 5.00405:
    HOME=/home/nick
    LANG=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/nick/bin:/usr/local/bin:/bin:/usr/bin:/usr/X11/bin:/usr/contrib/bin:/usr/games:/usr/sbin:/usr/ucb:/sbin:/usr/etc:/data3/src/emacs/bin/i386-unknown-bsdi2.1/
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

> $ perl5\.00405 \-T \-le 'delete $ENV\{PATH\}; my $d = $^X \. "" \. \`/bin/true\`'
> Insecure dependency in \`\` while running with \-T switch at \-e line 1\.
> $ perl5\.00405 \-T \-le 'delete $ENV\{PATH\}; my $d = "" \. $^X \. \`/bin/true\`'
> Insecure dependency in \`\` while running with \-T switch at \-e line 1\.
> $ perl5\.00405 \-T \-le 'delete $ENV\{PATH\}; my $d = lc \($^X\) \. \`/bin/true\`'
> Insecure dependency in \`\` while running with \-T switch at \-e line 1\.
> $ perl5\.00405 \-T \-le 'delete $ENV\{PATH\}; my $d = $^X \. \`/bin/true\`'
> $
> 
> I don't see why that last is any different\. Surely it should also be
> an insecure dependency?

Why ? Where's the unsafe operation ? You're concatenating tainted strings\.
None of the four examples should trigger this error\, IMO\.

**Comment by p5pRT** [comment]
### From @nwc10

On Thu\, May 22\, 2003 at 12&#8203;:04&#8203;:17AM \+0200\, Rafael Garcia\-Suarez wrote&#8203;:
> > $ perl5\.00405 \-T \-le 'delete $ENV\{PATH\}; my $d = $^X \. "" \. \`/bin/true\`'
> > Insecure dependency in \`\` while running with \-T switch at \-e line 1\.
> > $ perl5\.00405 \-T \-le 'delete $ENV\{PATH\}; my $d = "" \. $^X \. \`/bin/true\`'
> > Insecure dependency in \`\` while running with \-T switch at \-e line 1\.
> > $ perl5\.00405 \-T \-le 'delete $ENV\{PATH\}; my $d = lc \($^X\) \. \`/bin/true\`'
> > Insecure dependency in \`\` while running with \-T switch at \-e line 1\.
> > $ perl5\.00405 \-T \-le 'delete $ENV\{PATH\}; my $d = $^X \. \`/bin/true\`'
> > $
> > 
> > I don't see why that last is any different\. Surely it should also be
> > an insecure dependency?
> 
> Why ? Where's the unsafe operation ? You're concatenating tainted strings\.
> None of the four examples should trigger this error\, IMO\.

Hmm yes\. \`/bin/true\` needs security before it runs\, but the \. happens
afterwards\. OK\, so why aren't the first 3 clean?

Nicholas Clark

**Comment by p5pRT** [comment]
### From perl5-porters@perl.org

On Wed\, 21 May 2003 23&#8203;:06&#8203;:25 \+0100\, Nicholas Clark wrote \(in part\)&#8203;:

nc> Hmm yes\. \`/bin/true\` needs security before it runs\, but the \. happens
nc> afterwards\. OK\, so why aren't the first 3 clean?

Because the first 3 have a C\<\.> that's already happened by the time we
evaluate the C\<qx\`\`>\, and thus the concatentation of tainted strings
during the current expression makes the expression already tainted\.  The
4th one hasn't done more than fetch a variable yet\, which gives us tainted
data on the \(perl\) stack\, but not a tainted expression\.  Remember the
words from L\<perlsec>&#8203;:

&nbsp;    This function makes use of the fact that the presence of tainted data
&nbsp;    anywhere within an expression renders the entire expression tainted\. It
&nbsp;    would be inefficient for every operator to test every argument for
&nbsp;    taintedness\. Instead\, the slightly more efficient and conservative
&nbsp;    approach is used that if any tainted value has been accessed within the
&nbsp;    same expression\, the whole expression is considered tainted\.

Basically\, the $^X just references the value\, but doesn't B\<access> it\.
The C\<\.> is what accesses it\.  Applying an OP\_STRINGIFY \(C\<"$^X">\) also
accesses it\, thus tainting the expression\.

&nbsp;    \-\-s\.

**Comment by p5pRT** [comment]
### From @schwern

> Because the first 3 have a C\<\.> that's already happened by the time we
> evaluate the C\<qx\`\`>\, and thus the concatentation of tainted strings
> during the current expression makes the expression already tainted\.  The
> 4th one hasn't done more than fetch a variable yet\, which gives us tainted
> data on the \(perl\) stack\, but not a tainted expression\.

So is the consensus "not a bug"?

**Comment by p5pRT** [comment]
@schwern - Status changed from 'open' to 'stalled'

**Comment by p5pRT** [comment]
### From @nwc10

On Fri\, Dec 10\, 2004 at 02&#8203;:42&#8203;:26AM \-0000\, Michael G Schwern via RT wrote&#8203;:
> > Because the first 3 have a C\<\.> that's already happened by the time we
> > evaluate the C\<qx\`\`>\, and thus the concatentation of tainted strings
> > during the current expression makes the expression already tainted\.  The
> > 4th one hasn't done more than fetch a variable yet\, which gives us tainted
> > data on the \(perl\) stack\, but not a tainted expression\.
> 
> So is the consensus "not a bug"?

It's documented behaviour\. The behaviour of tainting is documented as being
erratic\. \(Strictly\, as having the option of erring on the side of paranoia
if it wants\. But couple that with tainting bugs\, and I don't consider
"erratic" to be slander\)

:\-\(

I think that it can be closed \- I added a todo wishlist item about all this
in pod/perltodo\.pod

Nicholas Clark

**Comment by p5pRT** [comment]
### From @jkeenan

On Fri Dec 10 02&#8203;:23&#8203;:41 2004\, nicholas wrote&#8203;:
> On Fri\, Dec 10\, 2004 at 02&#8203;:42&#8203;:26AM \-0000\, Michael G Schwern via RT
> wrote&#8203;:
> > > Because the first 3 have a C\<\.> that's already happened by the
> time we
> > > evaluate the C\<qx\`\`>\, and thus the concatentation of tainted
> strings
> > > during the current expression makes the expression already
> tainted\.  The
> > > 4th one hasn't done more than fetch a variable yet\, which gives us
> tainted
> > > data on the \(perl\) stack\, but not a tainted expression\.
> >
> > So is the consensus "not a bug"?
> 
> It's documented behaviour\. The behaviour of tainting is documented as
> being
> erratic\. \(Strictly\, as having the option of erring on the side of
> paranoia
> if it wants\. But couple that with tainting bugs\, and I don't consider
> "erratic" to be slander\)
> 
> :\-\(
> 
> I think that it can be closed \- I added a todo wishlist item about all
> this
> in pod/perltodo\.pod
> 
> Nicholas Clark
> 


This ticket has been in the Stalled queue since 2004\.  Can it be closed?

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'stalled' to 'open'

**Comment by p5pRT** [comment]
### From @jkeenan

On Sun Oct 07 07&#8203;:16&#8203;:47 2012\, jkeenan wrote&#8203;:
> On Fri Dec 10 02&#8203;:23&#8203;:41 2004\, nicholas wrote&#8203;:
> > On Fri\, Dec 10\, 2004 at 02&#8203;:42&#8203;:26AM \-0000\, Michael G Schwern via RT
> > wrote&#8203;:
> > > > Because the first 3 have a C\<\.> that's already happened by the
> > time we
> > > > evaluate the C\<qx\`\`>\, and thus the concatentation of tainted
> > strings
> > > > during the current expression makes the expression already
> > tainted\.  The
> > > > 4th one hasn't done more than fetch a variable yet\, which gives us
> > tainted
> > > > data on the \(perl\) stack\, but not a tainted expression\.
> > >
> > > So is the consensus "not a bug"?
> > 
> > It's documented behaviour\. The behaviour of tainting is documented as
> > being
> > erratic\. \(Strictly\, as having the option of erring on the side of
> > paranoia
> > if it wants\. But couple that with tainting bugs\, and I don't consider
> > "erratic" to be slander\)
> > 
> > :\-\(
> > 
> > I think that it can be closed \- I added a todo wishlist item about all
> > this
> > in pod/perltodo\.pod
> > 
> > Nicholas Clark
> > 
> 
> 
> This ticket has been in the Stalled queue since 2004\.  Can it be closed?
> 
> Thank you very much\.
> Jim Keenan

No one spoke up for keeping this ticket open\.  I am closing it\.  Thank
you very much\.

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'resolved'
